### PR TITLE
azure: support secondary ipv6 addresses

### PR DESCRIPTION
**Pull request by blackboxsw**
Azure's Instance Metadata Service (IMDS) reports multiple IPv6
addresses, via the http://169.254.169.254/metadata/instance/network
route.

When the the list of ipAddresses under the 'ipv6' interface key is
greater than one, the first IP is obtained via a dhcp6 configurtion,
and additional ipAddress entries will be configured with a static IPv6
CIDR.

### CODE REVIEW

File: cloudinit/sources/DataSourceAzure.py [hunk 1]
```diff
@@ -1345,12 +1345,21 @@ def parse_network_config(imds_metadata):
                 for addr6 in intf['ipv6']['ipAddress']:
                     privateIpv6 = addr6['privateIpAddress']
                     if privateIpv6:
-                        dev_config['dhcp6'] = True
-                        # non-primary interfaces should have a higher
-                        # route-metric (cost) so default routes prefer
-                        # primary nic due to lower route-metric value
-                        dev_config['dhcp6-overrides'] = dhcp_override
-                        break
+                        if dev_config.get('dhcp6', False):
```
Comment (raharper) [inline cloudinit/sources/DataSourceAzure.py]: What do you think about enumeration so we know we're on the primary interface.

Also,  if we are on a single interface, we need to dhcp6 *and* also assign ipv6 static addresses.

netplan already supports this:

```
ethernets:
   eth0:
      dhcp6: True
      addresses: [addr6, addr6, addr6]

```

File: cloudinit/sources/DataSourceAzure.py [hunk 2]
```diff
@@ -1345,12 +1345,21 @@ def parse_network_config(imds_metadata):
...
+                            # Append static address config for ip > 1
+                            prefix_v6 = intf['ipv6']['subnet'][0].get(
+                                'prefix', '128')
+                            if not dev_config.get('addresses'):
+                                dev_config['addresses'] = []
+                            dev_config['addresses'].append(
+                                '{ip}/{prefix}'.format(
+                                    ip=privateIpv6, prefix=prefix_v6))
```
Comment (raharper) [inline cloudinit/sources/DataSourceAzure.py]: I think we can reorganize the loop knowing that we will always just append additional IP addresses (v4 or v6).
Our default mode is:

```
dhcp4 = true, dhcp6 = False
if we have v6:
   dhcp6 = true
# ignore the first ip addr in each list, those are the
# values that will be obtained via DHCP
for ipaddr in inft['ipv4'][:1] + intf['v6'][1:]:
    addresses.append("{ip/prefix}")
dev_config['addresses'] = addresses

```

Comment (blackboxsw) [reply cloudinit/sources/DataSourceAzure.py]: I like this, it drops some lines of code too and have pushed a change to align these loops.

File: cloudinit/sources/DataSourceAzure.py [hunk 3]
```diff
@@ -1324,42 +1324,33 @@ def parse_network_config(imds_metadata):
...
-                                '{ip}/{prefix}'.format(
-                                    ip=privateIpv6, prefix=prefix_v6))
+                            default_prefix = '24'
                         else:
                             dev_config['dhcp6'] = True
                             # non-primary interfaces should have a higher
                             # route-metric (cost) so default routes prefer
                             # primary nic due to lower route-metric value
                             dev_config['dhcp6-overrides'] = dhcp_override
+                            default_prefix = '128'
```
Comment (raharper) [inline cloudinit/sources/DataSourceAzure.py]: Let's add a comment around the slicing

The first IP is obtained via DHCP, we add all additional IPs as static addresses.

File: cloudinit/sources/DataSourceAzure.py (hunk 1 above)
Comment (blackboxsw) [reply cloudinit/sources/DataSourceAzure.py]: I'm not sure I understand your comment here. Is the example netplan for a nic with 4 total ipv6 addresses listed by IMDS or 3 total?

Currently for a nic with 2 ipv4 and 3 ipv6 addrs in IMDS you'd get

ethernets:
  eth0:
   dhcp6: True
   dhcp4: True
   addresses: [addr4#2/prefix4, addr6#2/prefix6, addr6#3/prefix6, addr6#4/prefix6]

Comment (raharper) [reply cloudinit/sources/DataSourceAzure.py]: Sorry, it was incomplete, I just meant that we can extract/combine all of the v4/v6 static ips beyond the first IP on each of v4/v6  which obtain IP via DHCP.

The code you pushed does exactly what I was thinking.

File: cloudinit/sources/DataSourceAzure.py [hunk 4]
```diff
@@ -1321,6 +1321,9 @@ def parse_network_config(imds_metadata):
             LOG.debug('Azure: generating network configuration from IMDS')
             network_metadata = imds_metadata['network']
             for idx, intf in enumerate(network_metadata['interface']):
+                # First IPv4 and/or IPv6 address will be obtained via DHCP.
+                # Any additional IPs of each type will be set as static
+                # addresses with a route matching the subnet prefix.
```
Comment (raharper) [inline cloudinit/sources/DataSourceAzure.py]: Let's drop the 'with a route matching the subnet prefix'  Right? Aren't we just appending the addresses and netmask cidr value?

Comment (blackboxsw) [reply cloudinit/sources/DataSourceAzure.py]: We are appending the netmask CIDR prefix from subnet. Doesn't that add a route config entry as a result?

Comment (raharper) [reply cloudinit/sources/DataSourceAzure.py]: tl;dr  we're not emitting static routes for additional IPs within a subnet (and we don't need to)

1)  the ips are per subnet, which has a fixed CIDR, so they won't add any additional route entry just for an additional IP address that's in the same subnet
2) even if we did have additional subnets outside of the primary interface subnet, the code does not emit any route specific metric values; 
3) if we needed to explicitly create a higher metric subnet route for a particular IP, then we need to emit static route configuration in netplan which we are not doing at this time

Comment (raharper) [reply cloudinit/sources/DataSourceAzure.py]: the language suggests that we're emitting a route entry; where what actually happens is when the static ip is added, the kernel will create a subnet route by default if one does not exist; we don't control this exclusively and in particular, we're certainly not setting the metric on it.

Comment (blackboxsw) [reply cloudinit/sources/DataSourceAzure.py]: +1 @raharper.  Makes sense and dropping the comment

**Pull request merged**